File: src/insertionSort/insertionSort.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

###################################################
#                       					      #
# A simple sorting program using Insertion Sort   #
# and vectors in Python 					      #
# Author: Jesús Iván Gastelum Romero              #
# Contributors: Luis Edmundo Espinoza Larios      #
#                       					      #
###################################################

from time import time


def insertionSort(v):
	for i in range(1,len(v)):
		tmp = v[i]
		j=i
		while j > 0 and (tmp < v[j - 1]):
			v[j] = v[j-1]
			j = j - 1
		v[j] = tmp
	print '\nInsertion Sort:\n ', v

#*-----------------------------*#
#         MAIN PROGRAM          #
#*-----------------------------*#

start_time = time()
VECTOR = [7 , -5 , 0, 2 , 16, -9, 21, -874, 0, 8, 16, -210]
insertionSort(VECTOR)
end_time = time() - start_time
print '\nExecution Time: ', end_time, '\n'

File: src/insertionSort/insertionSort.c
/*
 * An implementation of the insertion sort algorithm
 * Author Luis Edmundo Espinoza Larios
 * Contributors Jesús Iván Gastelum Romero
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>
#include "util.h"

/*
 * Sorts a given array using insertion sort algorithm
 */
void sort(int[]);

int main()
{
    static int array[SIZE];
    struct timeval start, end;

    srand(time(NULL));

    generateArray(array);

    gettimeofday(&start, NULL);

    sort(array);
    printArray(array);

    gettimeofday(&end, NULL);

    printf ("Execution time: %f us\n", (double) (end.tv_usec - start.tv_usec) / 1000000 + (double) (end.tv_sec - start.tv_sec));

    return 0;
}

void sort(int array[])
{
    int i, j, temp;

    for (i = 1; i < SIZE; i++) {
        j = i;

        while (j > 0 && array[j - 1] > array[j]) {
            temp = array[j];
            array[j] = array[j - 1];
            array[j - 1] = temp;
            j -= 1;
        }
    }
}
